### PR TITLE
Use registerBinding in test suite plugin

### PR DESCRIPTION
**Pull request by big-guy**
refs #18260

### CODE REVIEW

**Comment by big-guy** [comment]
@bot-gradle test this

File: subprojects/plugins/src/main/java/org/gradle/api/plugins/JvmTestSuitePlugin.java [hunk 1]
```diff
@@ -49,8 +55,8 @@ public void apply(Project project) {
         JavaPluginExtension java = project.getExtensions().getByType(JavaPluginExtension.class);
         TestingExtension testing = project.getExtensions().getByType(TestingExtension.class);
         ExtensiblePolymorphicDomainObjectContainer<TestSuite> testSuites = testing.getSuites();
-        testSuites.registerFactory(TestSuite.class, name -> project.getObjects().newInstance(DefaultJvmTestSuite.class, name, project, java));
-        testSuites.registerFactory(JvmTestSuite.class, name -> project.getObjects().newInstance(DefaultJvmTestSuite.class, name, project, java));
+        testSuites.registerBinding(TestSuite.class, DefaultJvmTestSuite.class);
```
Comment (tresat) [inline subprojects/plugins/src/main/java/org/gradle/api/plugins/JvmTestSuitePlugin.java:58]: I see in the Javadoc that this is preferable, but why?  What does it add?  How should I understand where the "name" and "java" arguments are injected from?

Comment (tresat) [inline subprojects/plugins/src/main/java/org/gradle/api/plugins/JvmTestSuitePlugin.java:58]: I see that the javadoc says to prefer this method.  But how does the "name" argument get injected here?  Is that just understood to be the name of the block?

**Review by tresat — APPROVED** [review]

File: subprojects/plugins/src/main/java/org/gradle/api/plugins/JvmTestSuitePlugin.java (hunk 1 above)
Comment (big-guy) [reply subprojects/plugins/src/main/java/org/gradle/api/plugins/JvmTestSuitePlugin.java:58]: I can't find any javadoc that explains this, but by using registerBinding, Gradle is responsible for instantiating the element completely. When someone uses registerFactory, they could create elements by using `new` and lose all of our runtime decoration. Elements created with `new` and not ObjectFactory have slightly different behavior when it comes to Action/Closure methods and Property methods. 

With Gradle responsible for instantiating the element, we require that the constructor has a String as the first parameter. No other parameters can be provided except injectable services (which services are available depends on which scope we're in, but it's usually anything available to a project). We mention this, but we could be a lot clearer:
https://docs.gradle.org/current/userguide/custom_gradle_types.html#nameddomainobjectcontainer

Under the covers, when someone calls create for an element, we know the name then and we use a NamedDomainObjectFactory that uses ObjectFactory (or our internal Instantiator equivalent) to create the instance and pass the name given to us.

That's happening here: https://github.com/gradle/gradle/blob/9b631be71b9fa46ffd180b082f48b89a0fe9f91b/subprojects/core/src/main/java/org/gradle/api/internal/DefaultPolymorphicDomainObjectContainer.java#L105-L109

The lookup from type -> factory is here:
https://github.com/gradle/gradle/blob/9b631be71b9fa46ffd180b082f48b89a0fe9f91b/subprojects/core/src/main/java/org/gradle/api/internal/DefaultPolymorphicDomainObjectContainer.java#L63
https://github.com/gradle/gradle/blob/7302bd21010533bf0eff545479b2ffd503370dfb/subprojects/core/src/main/java/org/gradle/api/internal/DefaultPolymorphicNamedEntityInstantiator.java#L45-L51

**Comment by big-guy** [comment]
@bot-gradle test and merge

**Comment by big-guy** [comment]
Thanks for the review @tresat

File: subprojects/plugins/src/main/java/org/gradle/api/plugins/JvmTestSuitePlugin.java (hunk 1 above)
Comment (tresat) [reply subprojects/plugins/src/main/java/org/gradle/api/plugins/JvmTestSuitePlugin.java:58]: Thanks for the write up!

**Pull request merged**